### PR TITLE
Add @typescript-eslint/recommended-requiring-type-checking to eslint ruleset

### PR DESCRIPTION
**Pull request by mattseddon**
The main purpose of this PR is to turn on the following rule: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/await-thenable.md (follow up from [this](https://github.com/iterative/vscode-dvc/pull/3053#discussion_r1062053868)). 

The above rule requires [type information](https://typescript-eslint.io/linting/typed-linting) to run.

I thought that it would be a good time to turn on [plugin:@typescript-eslint/recommended-requiring-type-checking](https://typescript-eslint.io/#recommended-requiring-type-checking)

Most of the changes are to appease [no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-floating-promises.md) which was already on but not working.

### CODE REVIEW

File: extension/src/processExecution.ts [hunk 1]
```diff
@@ -63,10 +63,10 @@ export const executeProcess = async (
 export const processExists = (pid: number): Promise<boolean> =>
   doesProcessExists(pid)
 
-export const stopProcesses = (pids: number[]): boolean => {
+export const stopProcesses = async (pids: number[]): Promise<boolean> => {
   let allKilled = true
   for (const pid of pids) {
-    if (!processExists(pid)) {
+    if (!(await processExists(pid))) {
```
Comment (mattseddon) [inline extension/src/processExecution.ts:69]: [F] actual bug

File: extension/src/telemetry/index.ts [hunk 2]
```diff
@@ -55,7 +55,9 @@ const sanitizeProperty = (
     sanitizedProperties[key] = convertProperty(value)
   } catch (error: unknown) {
     Logger.error(
-      `Failed to serialize ${key} for ${String(eventName)}: ${error}`
+      `Failed to serialize ${key} for ${String(eventName)}: ${JSON.stringify(
```
Comment (mattseddon) [inline extension/src/telemetry/index.ts:58]: [F] Another bug

File: extension/src/plots/vega/util.ts [hunk 3]
```diff
@@ -1,3 +1,8 @@
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable @typescript-eslint/no-unsafe-return */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
Comment (mattseddon) [inline extension/src/plots/vega/util.ts:5]: [F] This is in the too hard basket for this PR

File: extension/src/test/e2e/util.ts [hunk 4]
```diff
@@ -113,14 +113,16 @@ export const deleteAllExistingExperiments = async () => {
...
+    deleteNonWorkspaceExperiments.elem.isDisplayed()
+  )
   await browser.keys('Enter')
 
   const deleteAllNonTagExperiments = await workbench.executeCommand(
     'DVC: Garbage Collect Experiments'
   )
 
-  browser.waitUntil(() => deleteAllNonTagExperiments.elem.isDisplayed())
+  await browser.waitUntil(() => deleteAllNonTagExperiments.elem.isDisplayed())
```
Comment (mattseddon) [inline extension/src/test/e2e/util.ts:125]: [F] Two bugs

File: extension/src/experiments/workspace.ts [hunk 5]
```diff
@@ -187,11 +187,13 @@ export class WorkspaceExperiments extends BaseWorkspaceWebviews<
     this.updatesPaused.fire(false)
   }
 
-  public getCwdThenReport(commandId: CommandId) {
-    const stdout = this.getCwdThenRun(commandId)
-    if (!stdout) {
```
Comment (mattseddon) [inline extension/src/experiments/workspace.ts:192]: [F] This was another bug (Promise is always truthy)

File: extension/src/test/cli/util.ts [hunk 6]
```diff
@@ -20,6 +20,7 @@ const gitExecutor = new GitExecutor()
 
 let demoInitialized: Promise<string>
 export const initializeDemoRepo = (): Promise<string> => {
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
   if (!demoInitialized) {
```
Comment (mattseddon) [inline extension/src/test/cli/util.ts:24]: this is the correct check (not undefined)

**Review by sroy3 — APPROVED** [review]
Seems like this was greatly benefitial in finding bugs and probably will be in the future as well. Good job!

Should we turn `'@typescript-eslint/no-unsafe-return'` and the likes off for the webview? Seems like we're disabling the rule a lot.

**Review by julieg18 — APPROVED** [review]
Great work!

**Comment by mattseddon** [comment]
> Seems like this was greatly benefitial in finding bugs and probably will be in the future as well. Good job!
> 
> Should we turn `'@typescript-eslint/no-unsafe-return'` and the likes off for the webview? Seems like we're disabling the rule a lot.

I spent a really long time getting the new rules turned on and fixing the commented-out lines in the webview would have pushed me over the edge. `'@typescript-eslint/no-unsafe-return'` is a good rule and I think we should use it. If we turn it on now then we can avoid any further breakages going forwards. Hopefully, we can slowly remove the comments over time (haha).

**Pull request merged**